src/main/java/medic/gateway/alert/SmsSender.java

```diff
@@ -114,22 +119,23 @@ private void sendSms_dummy(WoMessage m) {
 		db.updateStatus(m, SENT, DELIVERED);
 	}
 
-	private ArrayList<PendingIntent> intentsFor(String intentType, WoMessage m, ArrayList<String> parts) {
+	private ArrayList<PendingIntent> intentsFor(String intentType, WoMessage m, int mIdx, ArrayList<String> parts, Class<?> receiverClass) {
 		int totalParts = parts.size();
 		ArrayList<PendingIntent> intents = new ArrayList<>(totalParts);
 		for(int partIndex=0; partIndex<totalParts; ++partIndex) {
-			intents.add(intentFor(intentType, m, partIndex, totalParts));
+			intents.add(intentFor(intentType, m, mIdx, partIndex, totalParts, receiverClass));
 		}
 		return intents;
 	}
 
-	private PendingIntent intentFor(String intentType, WoMessage m, int partIndex, int totalParts) {
-		Intent intent = new Intent(intentType);
+	private PendingIntent intentFor(String intentType, WoMessage m, int mIdx, int partIndex, int totalParts, Class<?> receiverClass) {
+		Intent intent = new Intent(ctx, receiverClass);
+		intent.setAction(intentType);
 		intent.putExtra("id", m.id);
 		intent.putExtra("partIndex", partIndex);
 		intent.putExtra("totalParts", totalParts);
 
-		return PendingIntent.getBroadcast(ctx, UNUSED_REQUEST_CODE, intent, PendingIntent.FLAG_ONE_SHOT);
+		return PendingIntent.getBroadcast(ctx, mIdx, intent, PendingIntent.FLAG_ONE_SHOT);
```

Comment:
Great find! Using the same requestCode for each Message is a terrible idea!

I'm not sure about using the `mIdx` for the `requestCode`. To be certain we'd have to check the threading of these calls, because if the `SmsSender` is called again the index will start at 0 and potentially be duplicated again. Furthermore, if the same message is in a different position in the array then then it may get sent twice. It would be better to use a unique integer based on a property of the Message so it's deduped against sending twice, but doesn't dedupe other messages.

For this I think you could use the `m.id.hashCode()` or something? The `hashCode` may collide but it's unlikely, and the message will be retried later so the occasional collision isn't the end of the world...

```suggestion
		return PendingIntent.getBroadcast(ctx, m.id.hashCode(), intent, PendingIntent.FLAG_ONE_SHOT);
```

This also means you don't need to pass the `mIdx` all the way down the stack! Even better, if we ever want to fetch the intent again we can get it by just hashing the `m.id` again.

Reply:
That's good idea, I was thinking on adding a new property in the message class, but was worried about the integer size limitation.. the hashCode should be enough. Thanks!